internal/dkg/actions.go

```diff
@@ -1,60 +1,91 @@
-) error {
-	return d.executeActionWithCallback(name, beaconID, action, nil)
-}
-
-// executeActionWithCallback fetches the latest DKG state, applies the action to it, passes that new state
-// to a callback then stores the new state in the database if the callback was successful
-func (d *Process) executeActionWithCallback(
-	name string,
-	beaconID string,
-	createNewState func(me *drand.Participant, current *DBState) (*DBState, error),
-	callback func(me *drand.Participant, newState *DBState) error,
-) error {
-	var err error
-	d.log.Infow(fmt.Sprintf("Processing %s", name), "beaconID", beaconID)
+// gossip marks a DKG packet as seen and sends it to the other parties in the network (that are passed in)
+//
+//nolint:gocritic // ewww the linter wants me to use named parameters
+func (d *Process) gossip(
+	me *drand.Participant,
+	recipients []*drand.Participant,
+	packet *drand.GossipPacket,
+) (chan bool, chan error) {
+	done := make(chan bool, 1)
+	errChan := make(chan error, 1)
 
-	defer func() {
-		if err != nil {
-			d.log.Errorw(fmt.Sprintf("Error processing %s", name), "beaconID", beaconID, "error", err)
-		} else {
-			d.log.Infow(fmt.Sprintf("%s successful", name), "beaconID", beaconID)
-		}
-	}()
+	if len(recipients) == 0 {
+		done <- true
+		return done, errChan
+	}
 
-	me, err := d.identityForBeacon(beaconID)
-	if err != nil {
-		return err
+	if packet.Metadata == nil {
+		errChan <- errors.New("cannot process a packet without metadata")
+		return done, errChan
 	}
 
-	current, err := d.store.GetCurrent(beaconID)
-	if err != nil {
-		return err
+	// add the packet to the SeenPackets set,
+	// so we don't try and reprocess it when it gets gossiped back to us!
+	packetSig := hex.EncodeToString(packet.Metadata.Signature)
+	d.SeenPackets[packetSig] = true
+
+	wg := sync.WaitGroup{}
+
+	// we don't want to gossip to our own node
+	if util.Contains(recipients, me) {
+		wg.Add(len(recipients) - 1)
+	} else {
+		wg.Add(len(recipients))
 	}
+	for _, participant := range recipients {
+		p := participant
+		if p.Address == me.Address {
+			continue
+		}
 
-	nextState, err := createNewState(me, current)
-	if err != nil {
-		return err
+		// attempt to gossip with exponential backoff
+		go func() {
+			err := sendToPeer(d.internalClient, p, packet)
+			if err != nil {
+				d.log.Warnw("tried gossiping a packet but failed", "addr", p.Address, "packet", packetSig[0:8], "err", err)
+				errChan <- err
+			}
+			wg.Done()
+		}()
 	}
 
-	if callback != nil {
-		err = callback(me, nextState)
-		if err != nil {
-			return err
+	// signal the done channel when all the gossips + retries have been completed
+	go func() {
+		wg.Wait()
+		done <- true
+	}()
+
+	return done, errChan
+}
+
+func sendToPeer(client net.DKGClient, p *drand.Participant, packet *drand.GossipPacket) error {
+	retries := 8
+	backoff := 250 * time.Millisecond
+
+	peer := net.CreatePeer(p.Address, p.Tls)
+	var err error
+	for i := 0; i < retries; i++ {
```

Comment:
Shouldn't this have some kind of context and select to cancel early in case it got canceled from somewhere else?

Reply:
you can even do :

``` 
select {
	case <- ctx.Done():
		return fmt.Errorf("ctx canceled")
	case <- time.After(time.Duration(i+1) * backoff):
}
``` 
or something like that at the end